src/components/PurchaseModal/SwapAndBuy.tsx

```diff
@@ -0,0 +1,310 @@
+  const { data: walletClient } = useWalletClient()
+  const publicClient = usePublicClient()
+  const [swapsInProgress, setSwapsInProgress] = useState<string[]>([])
+
+  const { data: tokenSaleDetailsData, isLoading: tokenSaleDetailsDataIsLoading } = useReadContract({
+    abi: SALES_CONTRACT_ABI,
+    functionName: 'tokenSaleDetails',
+    chainId: CHAIN_ID,
+    address: SALES_CONTRACT_ADDRESS,
+    args: [BigInt(args.tokenId)]
+  })
+
+  const {
+    data: allowanceData,
+    isLoading: allowanceIsLoading,
+    refetch: refechAllowance
+  } = useReadContract({
+    abi: ERC_20_CONTRACT_ABI,
+    functionName: 'allowance',
+    chainId: CHAIN_ID,
+    address: (currencyData?.address || zeroAddress) as Hex,
+    args: [userAddress, SALES_CONTRACT_ADDRESS],
+    query: {
+      enabled: !!currencyData && !!userAddress
+    }
+  })
+
+  const price = (tokenSaleDetailsData as TokenSaleDetailsData)?.cost || 0n
+  const isApproved: boolean = (allowanceData as bigint) >= BigInt(price)
+
+  const { data: swapQuotes, isLoading: swapQuotesIsLoading } = useSwapQuotes({
+    userAddress: userAddress ?? '',
+    currencyAddress: currencyData?.address || '',
+    chainId: args.chainId,
+    currencyAmount: price.toString(),
+    withContractInfo: true
+  })
+
+  useEffect(() => {
+    clearCachedQuotes()
+    setSwapsInProgress([])
+  }, []) // eslint-disable-line react-hooks/exhaustive-deps
+
+  const isLoading: boolean = currencyIsLoading || tokenSaleDetailsDataIsLoading || allowanceIsLoading || swapQuotesIsLoading
+
+  const onClickPurchase = async (swapQuote: SwapQuotesWithCurrencyInfo) => {
+    if (!walletClient || !userAddress || !publicClient || !userAddress || !currencyData || !connector) {
+      return
+    }
+
+    const swapQuoteAddress = swapQuote.info?.address || ''
+
+    setSwapsInProgress([...swapsInProgress.filter(address => compareAddress(address, swapQuoteAddress)), swapQuoteAddress])
+
+    try {
+      const walletClientChainId = await walletClient.getChainId()
+      if (walletClientChainId !== args.chainId) {
+        await walletClient.switchChain({ id: args.chainId })
+      }
+
+      const approveTxData = encodeFunctionData({
+        abi: ERC_20_CONTRACT_ABI,
+        functionName: 'approve',
+        args: [SALES_CONTRACT_ADDRESS, price]
+      })
+
+      /*   **
+       * Mint tokens.
+       * @param to Address to mint tokens to.
+       * @param tokenIds Token IDs to mint.
+       * @param amounts Amounts of tokens to mint.
+       * @param data Data to pass if receiver is contract.
+       * @param expectedPaymentToken ERC20 token address to accept payment in. address(0) indicates ETH.
+       * @param maxTotal Maximum amount of payment tokens.
+       * @param proof Merkle proof for allowlist minting.
+       * @notice Sale must be active for all tokens.
+       * @dev tokenIds must be sorted ascending without duplicates.
+       * @dev An empty proof is supplied when no proof is required.
+       */
```

Comment:
Remove?

Reply:
remove in commit [093807f](https://github.com/0xsequence/currency-abstraction-demo/pull/4/commits/093807f1eb0ffde78c2bd1a1cbfc8216d961db59)